### PR TITLE
lib/modules: improve error-message for undeclared options if prefix contains no options

### PR DESCRIPTION
**Pull request by Ma27**
###### Motivation for this change

An easy-to-make mistake when declaring e.g. a submodule is the accidental
confusion of `options` and `config`:

    types.submodule {
      config = {
        foo = mkOption { /* ... */ };
      };
    }

However the error-message

```
  The option `[definition 1-entry 1].foo' defined in `<expr.nix>' does not exist.
```

is fairly unhelpful because it seems as the options are declared at the
first sight. In fact, it took a colleague and me a while to track down such
a mistake a few days ago and we both agreed that this should be somehow caught
to save the time we spent debugging the module in question.

At first I decided to catch this error in the `submodules`-type directly
by checking whether `options` is undeclared, however this becomes fairly
complicated as soon as a submodule-declaration e.g. depends on existing
`config`-values which would've lead to some ugly `builtins.tryExec`-heuristic.

This patch now simply checks if the option's prefix has any options
defined if a point in evaluation is reached where it's clear that the
option in question doesn't exist. This means that this patch doesn't
change the logic of the module system, it only provides a more detailed
error in certain cases:

```
  The option `[definition 1-entry 1].foo' defined in `<expr.nix>' does not exist.

  However it seems as there are no options defined in [definition 1-entry 1]. Are you sure you've
  declared your options properly? This happens if you e.g. declared your options in `types.submodule'
  under `config' rather than `options'.
```

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by jonringer — APPROVED** [review]

**Comment by Ma27** [comment]
Rebased onto master. @Infinisil as you authored those changes, could you please take a look? :)

File: lib/modules.nix [hunk 1]
```diff
@@ -115,8 +115,21 @@ rec {
 
       checkUnmatched =
         if config._module.check && config._module.freeformType == null && merged.unmatchedDefns != [] then
-          let inherit (head merged.unmatchedDefns) file prefix;
-          in throw "The option `${showOption prefix}' defined in `${file}' does not exist."
+          let
+            inherit (head merged.unmatchedDefns) file prefix;
+            filteredSet = builtins.removeAttrs (attrByPath (sublist 0 (length prefix - 1) prefix) {} options)
+              [ "_definedNames" "_module" ];
```
Comment (Infinisil) [inline lib/modules.nix:121]: What's the reason behind `attrByPath (sublist 0 (length prefix - 1) prefix) {} options`?

Also, no need to filter out `_definedNames` anymore

Comment (Ma27) [reply lib/modules.nix:121]: > What's the reason behind attrByPath (sublist 0 (length prefix - 1) prefix) {} options?

To retrieve the "parent" of the current option. I'd like to check whether there are any options defined.

> Also, no need to filter out _definedNames anymore

:+1:

Comment (Ma27) [reply lib/modules.nix]: @Infinisil updated the `removeAttrs` expression.

**Comment by Ma27** [comment]
@Infinisil anything else tbd? :)

File: lib/modules.nix [hunk 2]
```diff
@@ -115,8 +115,21 @@ rec {
...
         if config._module.check && config._module.freeformType == null && merged.unmatchedDefns != [] then
-          let inherit (head merged.unmatchedDefns) file prefix;
-          in throw "The option `${showOption prefix}' defined in `${file}' does not exist."
+          let
+            inherit (head merged.unmatchedDefns) file prefix;
+            filteredSet = builtins.removeAttrs (attrByPath (sublist 0 (length prefix - 1) prefix) {} options)
+              [ "_module" ];
+            baseMsg = "The option `${showOption (prefix)}' defined in `${file}' does not exist.";
+          in
+            if filteredSet == {}
```
Comment (Infinisil) [inline lib/modules.nix:124]: After testing this a bit, I'm noticing that the error can trigger in other scenarios too, e.g. in a standard NixOS module (not a submodule) that looks like
```nix
{
  options.foo = {};
  config.foo.bar = {};
}
```

In such a case, seeing the error might be more confusing.

So instead I suggest to change this condition to just check whether there are _any_ options at all:
```suggestion
            if attrNames options == [ "_module" ]
```

This still triggers if no options are defined in the `submodule`, but it won't trigger in the above example.

File: lib/modules.nix [hunk 3]
```diff
@@ -115,8 +115,21 @@ rec {
...
+              then throw ''
+                ${baseMsg}
+
+                However it seems as there are no options defined in ${showOption prefix}. Are you sure you've
+                declared your options properly? This happens if you e.g. declared your options in `types.submodule'
+                under `config' rather than `options'.
```
Comment (Infinisil) [inline lib/modules.nix:130]: From looking at this, I just noticed that in #82743 I introduced a mistake in the error message for missing options! I created https://github.com/NixOS/nixpkgs/pull/95718 to fix it again, which will unfortunately give a merge conflict again. Alternatively you could incorporate the commit from that PR into this one.

With that changed, I suggest to change the error message to
```suggestion
                However there are no options defined in ${showOption prefix}. Are you sure you've
                declared your options properly? This can happen if you declared your options in `types.submodule'
                under `config' rather than `options'.
```

(note that this should be `prefix` and not `firstDef.prefix` from my PR)

**Review by infinisil — CHANGES_REQUESTED** [review]

File: lib/modules.nix (hunk 2 above)
Comment (Ma27) [reply lib/modules.nix]: Good catch!

File: lib/modules.nix (hunk 3 above)
Comment (Ma27) [reply lib/modules.nix]: Ah nice :-)
Rebased and updated accordingly.

File: lib/modules.nix [hunk 4]
```diff
@@ -115,8 +115,19 @@ rec {
...
-          in throw "The option `${showOption (prefix ++ firstDef.prefix)}' defined in `${firstDef.file}' does not exist."
+          let
+            firstDef = head merged.unmatchedDefns;
+            baseMsg = "The option `${showOption (prefix ++ firstDef.prefix)}' defined in `${firstDef.file}' does not exist.";
+          in
+            if attrNames options == [ "_module" ]
+              then throw ''
+                ${baseMsg}
+
+                However it seems as there are no options defined in ${showOption (prefix ++ firstDef.prefix)}. Are you sure you've
```
Comment (Infinisil) [inline lib/modules.nix:126]: This should be just `showOption prefix` now.

I also still suggest rewording this to `However there are no options defined in ...`, (since that's indeed the case, it doesn't just seem that way :))

Comment (Ma27) [reply lib/modules.nix:126]: Oof forgot that, sorry!

Comment (Ma27) [reply lib/modules.nix]: It's changed now.

**Review by infinisil — APPROVED** [review]
